internal/pkg/manifests/receive/builder.go

```diff
@@ -0,0 +1,397 @@
+package receive
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/thanos-community/thanos-operator/internal/pkg/manifests"
+
+	"github.com/prometheus/prometheus/model/labels"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/utils/ptr"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// Name is the name of the Thanos Receive component.
+	Name = "thanos-receive"
+
+	// RouterComponentName is the name of the Thanos Receive router component.
+	RouterComponentName = "thanos-receive-router"
+	// RouterHTTPPortName is the name of the HTTP port for the Thanos Receive router component.
+	RouterHTTPPortName = "http"
+	// RouterHTTPPort is the port number for the HTTP port for the Thanos Receive router component.
+	RouterHTTPPort = 10902
+
+	// IngestComponentName is the name of the Thanos Receive ingester component.
+	IngestComponentName = "thanos-receive-ingester"
+	// IngestGRPCPortName is the name of the gRPC port for the Thanos Receive ingester component.
+	IngestGRPCPortName = "grpc"
+	// IngestGRPCPort is the port number for the gRPC port for the Thanos Receive ingester component.
+	IngestGRPCPort = 10901
+	// IngestHTTPPortName is the name of the HTTP port for the Thanos Receive ingester component.
+	IngestHTTPPortName = "http"
+	// IngestHTTPPort is the port number for the HTTP port for the Thanos Receive ingester component.
+	IngestHTTPPort = 10902
+	// IngestRemoteWritePortName is the name of the remote write port for the Thanos Receive ingester component.
+	IngestRemoteWritePortName = "remote-write"
+	// IngestRemoteWritePort is the port number for the remote write port for the Thanos Receive ingester component.
+	IngestRemoteWritePort = 19291
+
+	// HashringConfigKey is the key in the ConfigMap for the hashring configuration.
+	HashringConfigKey = "hashrings.json"
+)
+
+// IngesterOptions for Thanos Receive components
+type IngesterOptions struct {
+	manifests.Options
+	StorageSize    resource.Quantity
+	ObjStoreSecret corev1.SecretKeySelector
+	Retention      string
+}
+
+// RouterOptions for Thanos Receive router
+type RouterOptions struct {
+	manifests.Options
+}
+
+// HashringOptions for Thanos Receive hashring
+type HashringOptions struct {
+	manifests.Options
+	HashringSettings map[string]HashringMeta
+}
+
+// HashringMeta represents the metadata for a hashring.
+type HashringMeta struct {
+	OriginalName             string
+	Replicas                 int32
+	Tenants                  []string
+	TenantMatcherType        TenantMatcher
+	AssociatedEndpointSlices discoveryv1.EndpointSliceList
+}
+
+// Endpoint represents a single logical member of a hashring.
+type Endpoint struct {
+	// Address is the address of the endpoint.
+	Address string `json:"address"`
```

Comment:
Wonder if we can add some validation to these address fields

Reply:
yes, will take care of that in the follow up when we do the hashring work. just wanted to port over the thanos types here for visibility